src/sentry/integrations/source_code_management/repository.py

```diff
@@ -1,39 +1,98 @@
-    # dynamically given a search query
-    repo_search = False
 
+class RepositoryIntegration(IntegrationInstallation, ABC):
+    @property
+    def codeowners_locations(self) -> list[str] | None:
+        """
+        A list of possible locations for the CODEOWNERS file.
+        """
+        return None
+
+    @property
+    def repo_search(self) -> bool:
+        return True
+
+    @property
+    @abstractmethod
+    def integration_name(self) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_client(self) -> RepositoryClient:
+        """Returns the client for the integration. The client must be a subclass of RepositoryClient."""
+        raise NotImplementedError
+
+    @abstractmethod
     def source_url_matches(self, url: str) -> bool:
-        """Checks if the url matches the integration's source url."""
+        """Checks if the url matches the integration's source url. Used for stacktrace linking."""
         raise NotImplementedError
 
-    def format_source_url(self, repo: Repository, filepath: str, branch: str) -> str:
-        """Formats the source code url used for stack trace linking."""
+    @abstractmethod
+    def format_source_url(self, repo: Repository, filepath: str, branch: str | None) -> str:
+        """Formats the source code url used for stacktrace linking."""
         raise NotImplementedError
 
+    @abstractmethod
     def extract_branch_from_source_url(self, repo: Repository, url: str) -> str:
-        """Extracts the branch from the source code url."""
+        """Extracts the branch from the source code url. Used for stacktrace linking."""
         raise NotImplementedError
 
+    @abstractmethod
     def extract_source_path_from_source_url(self, repo: Repository, url: str) -> str:
-        """Extracts the source path from the source code url."""
+        """Extracts the source path from the source code url. Used for stacktrace linking."""
         raise NotImplementedError
 
-    def check_file(self, repo: Repository, filepath: str, branch: str) -> str | None:
+    @abstractmethod
+    def has_repo_access(self, repo: RpcRepository) -> bool:
+        """Used for migrating repositories. Checks if the installation has access to the repository."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_repositories(self, query: str | None = None) -> Sequence[dict[str, Any]]:
+        """
+        Get a list of available repositories for an installation
+
+        >>> def get_repositories(self):
+        >>>     return self.get_client().get_repositories()
+
+        return [{
+            'name': display_name,
+            'identifier': external_repo_id,
+        }]
+
+        The shape of the `identifier` should match the data
+        returned by the integration's
+        IntegrationRepositoryProvider.repository_external_slug()
+        """
+        raise NotImplementedError
+
+    def get_unmigratable_repositories(self) -> Collection[RpcRepository]:
```

Comment:
Can this be list or some other tighter type than collection?

Reply:
yes this can be a list